epinorm/norm.py

```diff
@@ -282,24 +397,63 @@ def _get_location_from_strain(self, record):
                 return token
 
     def _format_location(self, record):
+        """
+        basically we combine the information from original_record_location_description and the extracted information
+        of the strain. We also clean them (removing unwanted characters)
+
+        we return a dictionaly in json format with all the fields
+        {
+            country
+            areas
+        }
+        """
+
+        def clean_token(token):
+            token = token.replace("_", " ")
+            token = re.sub(r"\(.+\)", "", token) 
+            token = re.sub(r"[^\w\s-]+", "", token)
+            token = token.strip()
+            token = token.lower() # for easier comparison later
+            return token
+
         location = json.loads(record["original_record_location_description"])
-        extracted_location = record["extracted_location"]
-        if "area" not in location:
-            location["area"] = extracted_location
-        country = str(coalesce(location["country"], ""))
-        area = str(coalesce(location["area"], ""))
-        places = []
-        for place in area.split(","):
-            place = place.replace("_", " ")
-            place = re.sub(r"\(.+\)", "", place)
-            place = re.sub(r"[^\w\s-]+", "", place)
-            place = place.strip()
-            if not place or place == country:
-                continue
-            places.append(place)
-        if places:
-            return ", ".join(places) + ", " + country
-        return country
+
+        # format country to make comparison easier
+        location["country"] = clean_token(location["country"])
+
+        new_places = [] # combine areas and extrated location together in this list
+
+        # format areas
+        for place in location["areas"]:
+
+            cleaned_place = clean_token(place)
+            
+            if cleaned_place != "" and cleaned_place != location["country"] and \
+               cleaned_place not in new_places:
+                new_places.append(cleaned_place)
+
+
+        # format extracted field
+        if record["extracted_location"]:
+
+            cleaned_place = clean_token(record["extracted_location"])
+
+            if cleaned_place != "" and cleaned_place != location["country"] and \
+               cleaned_place not in new_places:
+                new_places.append(cleaned_place)
+
```

Comment:
To avoid code duplication, combine location["areas"] and record["extracted_location"] in a single array and apply the processing logic once.